### ISSUE TITLE
Need a notification when MongoCk is done

### ISSUE THREAD

**Issue by frochi42** [cloudyrock/mongock#261]
We are using mongoCk in a Spring environment, and our applications run in a kubernetes-cluster which run liveness and readiness probe. We would like that our application is considered to be UP only after the MongoCK migration (we don't want to have traffic in the middle of migration).
We recognised that MongoCk is running after the Application-ready message of Spring, so by default our application is considered to by UP too early (at least if migration is performed).
The question is: how can we manage that requirement? We considered writing an own HealthIndicator, but this would require a callback from MongoCK when done. As far we could see from the code, there isn't such a mechanism yet.
Any ideas how we could do that?

**Comment by dieppa** [comment]
Hello, can you show me how you are using Mongock? I specially interested in 
- You Spring boot application and the annotations your are providing to it
- Mongock's configuration or, if you are not using @EnableMongock, the way you are building it with the traditional builder approach

**Comment by frochi42** [comment]
Hi, thanks for your answer.
We have one single Config class that looks like

@Configuration
@ConditionalOnProperty(havingValue = "true", name = "spring.data.mongodb.run-changelog")
@Slf4j
public class MongockConfig {

    @Value("${spring.data.mongodb.changelogPackage}")
    private String baseScanPackageName;

    @Value("${spring.data.mongodb.database}")
    private String database;

    public MongockConfig() {
        log.info("MongockConfig support enabled");
    }
    @Bean
    public MongockSpring5.MongockApplicationRunner mongockApplicationRunner(ApplicationContext springContext,
                                                                            MongoTemplate mongoTemplate) {
        MongockConfiguration config = new MongockConfiguration();
        config.setIndexCreation(false);
        config.setChangeLogsScanPackage(Collections.singletonList(baseScanPackageName));

        MongoSync4Driver mongoSync4Driver = SpringDataMongo3Driver.withDefaultLock(mongoTemplate);
        mongoSync4Driver.setIndexCreation(false);
        mongoSync4Driver.disableTransaction();

        Document mongockLockIndex = Document.parse("{\"unique\":true,\"key\":{\"key\":1},\"name\":\"key_1\",\"ns\":\"" + database + ".mongockLock\"}");
        Document mongockChangeLogIndex = Document.parse("{\"unique\":true,\"key\":{\"executionId\":1,\"author\":1,\"changeId\":1},\"name\":\"mongockChangeLog_1\",\"ns\":\"" + database + ".mongockChangeLog\"}");

        mongoTemplate.save(mongockLockIndex, AbstractCollectionConfig.SYS_IDX_COLLECTION_NAME);
        mongoTemplate.save(mongockChangeLogIndex, AbstractCollectionConfig.SYS_IDX_COLLECTION_NAME);

        MongockSpring5.MongockApplicationRunner mongockApplicationRunner = MongockSpring5.builder()
                .setDriver(mongoSync4Driver)
                .addChangeLogsScanPackage(baseScanPackageName)
                .setSpringContext(springContext)
                .setConfig(config)
                .buildApplicationRunner();

        return mongockApplicationRunner;
    }
}

Besides that, we only have DatabaseChangeLog classes:@ChangeLog(order = "001")
@Slf4j
public class DatabaseChangelog {

    private static final int PAGE_SIZE = 200;

    @ChangeSet(order = "001", id = "addIdToConditions", author = "company")
    public void migrate001(@Qualifier MongockTemplate mongockTemplate) {
...

Does that answer your question?

**Comment by dieppa** [comment]
Hello, 

to address the problem you're having we are developing two mechanism:
- Spring events. Basically for application based on Spring
- Subscribers. Although it's intended for non-spring application, it would work for Spring application to, in case Spring events is not suitable for any reason. 

We expect to have this at the end of the week or the next one. Meanwhile, have you tried to use `mongockApplicationRunner` instead `MongockApplicationRunner`? That guarantees once the Spring Context has been fully initialised, Mongock has already executed the migration. Then you can add a listener for `ApplicationReadyEvent`.

That's not ideal, but I think it will work as a temporal solution.

**Comment by frochi42** [comment]
Hi,
that seems to be feasable. We have to anyway subscribe to ApplicationReadyEvent, so that should be fine,
One question: is that the only difference between MongockInitializingBeanRunner and MongockApplicationRunner?

**Comment by dieppa** [comment]
Good to hear that.

Yes, the only different between both is that the first implements the interface `InitializingBean` and the second one `ApplicationRunner`.

ApplicationRunner(as well as CommandLineRunner) is an interface when you want to run that code when the application is fully started. On the on the other hand, according to the bean's lifecycle, before making the bean ready to use, the method `afterPropertiesSet` from interface  `InitialazingBean` is run, and there is where the migration takes place. If the migration failed, the bean wouldn't be ready and the Spring startup wouldn't happen.

I hope that answer your question :)

**Comment by frochi42** [comment]
Hi Dieppa,
just wanted to inform me whether you already implemented such a notification mechanism?
In the meantime we have an additional use case for that: We need to implement a graceful shutdown mechanism. Concerning MongoCK that means that we must postpone the shutdown during the DB migration with MongoCK.
Thanks a lot!

**Comment by dieppa** [comment]
Hi @frochi42 ,

We actually have a BETA version(still not documented properly) with that. I hadn't told you as I wanted to provide the documentation.

The version is 4.2.1.BETA.
Try this.

1. Provide the spring `ApplicationEventPublisher` to the builder with `setEventPublisher` method **only if you are not using @EnableMongock**

2. Subscribe to the events: `SpringMigrationSuccessEvent` and/or `SpringMigrationFailureEvent` like:

```java
package com.github.cloudyrock.spring.v5;

import io.changock.runner.spring.util.events.SpringMigrationSuccessEvent;
import org.springframework.context.ApplicationListener;

public class MyMongockEventHandler implements ApplicationListener<SpringMigrationSuccessEvent> {
  @Override
  public void onApplicationEvent(SpringMigrationSuccessEvent event) {
    //TODO your logic
  }
}
```


Please take into account that for sure the SuccessEvent will be enhanced, which won't really break anything, but also may change some other stuff that will may affect you slightly.

**Comment by dieppa** [comment]
I would appreciate if you can send me your feedback to dev@cloudyrock.io. Anything you would add, issues, suggestions..everything is welcome :)

**Comment by frochi42** [comment]
Hi, thanks for that explanation.
In the meantime I already noticed that new BETA version, which I downloaded. I just wasn't able to find out (without Docu) how to subscribe. Now I know and I will try. Thank you.
Just one additional remark: Would it be possible to also issue a "migrationstarted" event? Thank you...

**Comment by dieppa** [comment]
You're welcome.

**Just one additional remark: Would it be possible to also issue a "migrationstarted" event?** -> Actually is a good idea, we'll add that.

**Comment by frochi42** [comment]
I'm a bit confused.

- You say "Provide the spring ApplicationEventPublisher to the builder with setEventPublisher method" (yes we are using the builder method).
What is _the spring ApplicationEventPublisher_ ? Where to get it from?

- That public class MyMongockEventHandler: Shouldn't that be an @Component? If not, how to register it?
BTW: did you add some unit tests that show the supposed usage?
Thanks a lot!

**Comment by dieppa** [comment]
- You say "Provide the spring ApplicationEventPublisher to the builder with setEventPublisher method" (yes we are using the builder method).
What is the spring ApplicationEventPublisher ? Where to get it from?
**If you are using Spring, you just need to add that in your bean method, like the following code...**
```java
    @Bean
    public MongockStandalone.Runner updateMongoTemplateSchema(ApplicationEventPublisher eventPublisher, 
                                                              MongoTemplate mongoTemplate) {
        //TODO your builder, setting the eventPublisher
    }
```


- That public class MyMongockEventHandler: Shouldn't that be an @component? If not, how to register it?
**Yes, I just provided that as sample code, but you need to register that with annotation or in your configuration as a bean**

BTW: did you add some unit tests that show the supposed usage?
**we actually have a separated project where you can see the supposed usage. Please check [this github project](https://github.com/cloudyrock/mongock-integration-tests/tree/master/mongock-4/mongock-spring-v5/mongock-spring5-springdata3-it)**

I hope this resolve your issue :)

**Comment by frochi42** [comment]
Hi,
good news: I was able to get your proposed changes to work! Thanks a lot!
Now I would wait for that "migration started" event....

**Comment by frochi42** [comment]
Unfortunately, after doing some more tests, I encountered a problem.
As I wrote at the very beginning of this thread, our actual requirement is that our Application should not be shutdown during a running migration - the reason is that in such a case it is possible that the DB is left in an inconsistent state, from which we cannot recover (except manual interaction).

That means, we must implement a graceful-shutdown mechanism (I don't think that mongock has an option for that - correct me if I should be wrong).

The idea is, that I have a @Component "MongockGracefulShutdown" that implements a @PreDestruct method. This method checks whether we are in the middle of a running migration. If yes, it would simply enter a loop, until the migration has finished. To detect whether a migration is running I would use the "Migration Started" and "Migration ended " events from MongoCK. So far so good. Normally that works.

But if the application receives a shutdown-event from outside (kubernetes) during a running migration, the MongoCk-Runner is killed before the "Migration ended" event was sent! That means, my MongockGracefulShutdown never notices that the migration has ended, and the application is never killed!

Do you have any idea how to solve this issue?

**Comment by dieppa** [comment]
Hello, I wasn't sure what you needed for, but it was in our roadmap.

I assume you are not able to use transactions, as that would be very handy in your situation. This is not really a Mongock problem, but a MongoDB limitation(before transactions).

For your case, Long time ago I wrote [this in the Mongock version 3 readme](https://github.com/cloudyrock/mongock/blob/master/community/README_V3.md#mongo-transaction-limitations).  It's not a solution, but some design suggestions that may help you.

On top of that, we have in our roadmap [recovery](https://www.mongock.io/recovery) and [undo](https://www.mongock.io/undo). 

As said, this is part of the pre-transactions MongoDB limitations, so there is nothing Mongock can guarantee, but it can be worked-around by applying a proper changeLog design plus the help of [recovery](https://www.mongock.io/recovery) and/or [undo](https://www.mongock.io/undo).


On the other hand, the started event will be probably release in 4.2.2.BETA this evening or tomorrow.

**Comment by dieppa** [comment]
Hello @frochi42 ,

just to let you know that in version 4.2.2.BETA we have added `SpringMigrationStartedEvent`.

Was the previous comment in some way useful?

**Comment by frochi42** [comment]
Good Morning,
yes, your comment was really useful. I learned from them that even if a ChangeLog finishes successfully, you cannot be sure that the SW running the changelog will always run on migrated data. The reason for that is that there might be an older version of the SW (not knowing about the changelog) running in parallel, which might convert data back in the old format.
That means, DB upgrade is not easy and can most probably not be performed in one run.
And from that it follows that my initial idea (making sure my app cannot be shutdown during a migration with MongoCk) is not useful at all. Yes, it would lower the probability for data inconsistencies, but it would never be zero.
Therefor it is obviously necessary (as you also pointed out on your explanations) to handle this situation in each and every ChangeLog. I would have liked to avoid that (as now I need to tell each and every team member how to do it) and would have prefered a centralized solution. But as a result from our discussions, this seems to be not possible.

BTW, thanks for having added that new event so quickly!

Regards, frochi42

**Comment by dieppa** [comment]
Exactly, to make it right it's not straight forward. You need to take into account several aspects, like the infrastructure you are using, microservices, if you system is expected to be high available, etc. However, MongoDB transactions are very useful in this scenario.  Just to let you know that, If you need further assistance, we also provides support plans.

Regards
